server/interceptor/interceptor.go

```diff
@@ -151,17 +137,69 @@ func getRPCInfo(req interface{}, fullMethod, quotaUser string) (*rpcInfo, error)
 	}, nil
 }
 
-// parseFullMethod returns the service and method names as separate strings, without a trailing
-// slash in either of them.
-func parseFullMethod(fullMethod string) (string, string, error) {
-	if !strings.HasPrefix(fullMethod, "/") {
-		return "", "", status.Errorf(codes.Internal, "fullMethod must begin with '/': %v", fullMethod)
+func getRequestInfo(req interface{}) (trillian.TreeType, bool, error) {
+	if ok, readonly := getAdminRequestInfo(req); ok {
+		return trillian.TreeType_UNKNOWN_TREE_TYPE, readonly, nil
+	}
+	if ok, readonly := getLogRequestInfo(req); ok {
+		return trillian.TreeType_LOG, readonly, nil
+	}
+	if ok, readonly := getMapRequestInfo(req); ok {
+		return trillian.TreeType_MAP, readonly, nil
+	}
+	return trillian.TreeType_UNKNOWN_TREE_TYPE, false, fmt.Errorf("unmapped request type: %T", req)
+}
+
+func getAdminRequestInfo(req interface{}) (bool, bool) {
+	isAdmin := true
+	readonly := false
+	switch req.(type) {
+	case *trillian.CreateTreeRequest,
+		*trillian.DeleteTreeRequest,
+		*trillian.UpdateTreeRequest:
```

Comment:
Personal preference, but I think these would be easier to read if the `isXxx` flag is set in the case arms, rather than being set by default then cleared in a default arm.  You could even get to use the rarely-spotted `fallthrough` keyword!:
```go
func getAdminRequestInfo(req interface{}) (bool, bool) {
	isAdmin := false
	readonly := false
	switch req.(type) {
	case *trillian.GetTreeRequest,
		*trillian.ListTreesRequest:
		readonly = true
		fallthrough
	case *trillian.CreateTreeRequest,
		*trillian.DeleteTreeRequest,
		*trillian.UpdateTreeRequest:
		isAdmin = true
	}
	return isAdmin, readonly
}
```